# Guided Project: SQL Window Functions for Northwind Traders

## Objective

Northwind Taders is an international gourmet food distributor. Management would like insights to make strategic decisions in several aspects of the business. The projects focus is on:

1. Evaluating employee performance to boost productivity,
2. Understanding product sales and category performance to optimize inventory and marketing strategies,
3. Analyzing sales growth to identify trends, monitor company progress, and make more accurate forecasts,
4. And evaluating customer purchase behavior to target high-value customers with promotional incentives.

## Load sql plugin and connect to Northwind postgres db

In [559]:
!pip3 install ipython-sql

In [560]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [561]:
import getpass
pw = getpass.getpass('Énter password')
%sql postgresql://postgres:{pw}@localhost:5432/northwind

Énter password ········


## Getting to Know the Data

![alt text](/Northwind%20ER.png)

### Tables, data types and number of columns

In [562]:
%%sql
  SELECT table_name, data_type,
         COUNT(column_name) AS num_of_columns
    FROM information_schema.columns
   WHERE table_schema = 'public'
GROUP BY table_name, data_type
ORDER BY table_name, data_type

 * postgresql://postgres:***@localhost:5432/northwind
36 rows affected.


[('categories', 'bytea', 1),
 ('categories', 'character varying', 1),
 ('categories', 'smallint', 1),
 ('categories', 'text', 1),
 ('customer_customer_demo', 'character varying', 2),
 ('customer_demographics', 'character varying', 1),
 ('customer_demographics', 'text', 1),
 ('customers', 'character varying', 11),
 ('employee_territories', 'character varying', 1),
 ('employee_territories', 'smallint', 1),
 ('employees', 'bytea', 1),
 ('employees', 'character varying', 12),
 ('employees', 'date', 2),
 ('employees', 'smallint', 2),
 ('employees', 'text', 1),
 ('order_details', 'real', 2),
 ('order_details', 'smallint', 3),
 ('orders', 'character varying', 7),
 ('orders', 'date', 3),
 ('orders', 'real', 1),
 ('orders', 'smallint', 3),
 ('products', 'character varying', 2),
 ('products', 'integer', 1),
 ('products', 'real', 1),
 ('products', 'smallint', 6),
 ('region', 'character varying', 1),
 ('region', 'smallint', 1),
 ('shippers', 'character varying', 2),
 ('shippers', 'smallint', 1),
 ('suppliers', 'character varying', 10),
 ('suppliers', 'smallint', 1),
 ('suppliers', 'text', 1),
 ('territories', 'character varying', 2),
 ('territories', 'smallint', 1),
 ('us_states', 'character varying', 3),
 ('us_states', 'smallint', 1)]

### Order Details

- Discount appears to be a percentage

In [563]:
%%sql
SELECT *
  FROM order_details
 LIMIT 10;

 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[(10248, 11, 14.0, 12, 0.0),
 (10248, 42, 9.8, 10, 0.0),
 (10248, 72, 34.8, 5, 0.0),
 (10249, 14, 18.6, 9, 0.0),
 (10249, 51, 42.4, 40, 0.0),
 (10250, 41, 7.7, 10, 0.0),
 (10250, 51, 42.4, 35, 0.15),
 (10250, 65, 16.8, 15, 0.15),
 (10251, 22, 16.8, 6, 0.05),
 (10251, 57, 15.6, 15, 0.05)]

In [564]:
%%sql
SELECT *
  FROM orders
LIMIT 3;

 * postgresql://postgres:***@localhost:5432/northwind
3 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France'),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany'),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil')]

In [565]:
%%sql
SELECT *
  FROM products
 LIMIT 3;

 * postgresql://postgres:***@localhost:5432/northwind
3 rows affected.


[(1, 'Chai', 8, 1, '10 boxes x 30 bags', 18.0, 39, 0, 10, 1),
 (2, 'Chang', 1, 1, '24 - 12 oz bottles', 19.0, 17, 40, 25, 1),
 (3, 'Aniseed Syrup', 1, 2, '12 - 550 ml bottles', 10.0, 13, 70, 25, 0)]

In [566]:
%%sql
SELECT *
  FROM employees
 LIMIT 3;

 * postgresql://postgres:***@localhost:5432/northwind
3 rows affected.


[(1, 'Davolio', 'Nancy', 'Sales Representative', 'Ms.', datetime.date(1948, 12, 8), datetime.date(1992, 5, 1), '507 - 20th Ave. E.\\nApt. 2A', 'Seattle', 'WA', '98122', 'USA', '(206) 555-9857', '5467', None, 'Education includes a BA in psychology from Colorado State University in 1970.  She also completed The Art of the Cold Call.  Nancy is a member of Toastmasters International.', 2, 'http://accweb/emmployees/davolio.bmp'),
 (2, 'Fuller', 'Andrew', 'Vice President, Sales', 'Dr.', datetime.date(1952, 2, 19), datetime.date(1992, 8, 14), '908 W. Capital Way', 'Tacoma', 'WA', '98401', 'USA', '(206) 555-9482', '3457', None, 'Andrew received his BTS commercial in 1974 and a Ph.D. in international marketing from the University of Dallas in 1981.  He is fluent in French and  ... (150 characters truncated) ... es in March 1993.  Andrew is a member of the Sales Management Roundtable, the Seattle Chamber of Commerce, and the Pacific Rim Importers Association.', None, 'http://accweb/emmployees/fuller.bmp'),
 (3, 'Leverling', 'Janet', 'Sales Representative', 'Ms.', datetime.date(1963, 8, 30), datetime.date(1992, 4, 1), '722 Moss Bay Blvd.', 'Kirkland', 'WA', '98033', 'USA', '(206) 555-3412', '3355', None, 'Janet has a BS degree in chemistry from Boston College (1984).  She has also completed a certificate program in food retailing management.  Janet was hired as a sales associate in 1991 and promoted to sales representative in February 1992.', 2, 'http://accweb/emmployees/leverling.bmp')]

### Type Error Occurred with the employees table
Employees table had a photo column which seemed to be causing an error with the table visualisation. Updated photo column in database to remove the BMP images from this column.

```Update employees SET photo = null;```

In [567]:
%%sql
SELECT *
  FROM information_schema.columns
 WHERE column_name = 'photo'
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
1 rows affected.


[('northwind', 'public', 'employees', 'photo', 15, None, 'YES', 'bytea', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'northwind', 'pg_catalog', 'bytea', None, None, None, None, '15', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES')]

In [568]:
%%sql
-- Combine orders and customers tables to get more detailed information about each order.

SELECT *
  FROM orders AS o
  JOIN customers AS c
    ON o.customer_id = c.customer_id 
 LIMIT 5;



 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 'VINET', 'Vins et alcools Chevalier', 'Paul Henriot', 'Accounting Manager', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', '26.47.15.10', '26.47.15.11'),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', 'TOMSP', 'Toms Spezialitäten', 'Karin Josephs', 'Marketing Manager', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', '0251-031259', '0251-035695'),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', 'HANAR', 'Hanari Carnes', 'Mario Pontes', 'Accounting Manager', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', '(21) 555-0091', '(21) 555-8765'),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France', 'VICTE', 'Victuailles en stock', 'Mary Saveley', 'Sales Agent', '2, rue du Commerce', 'Lyon', None, '69004', 'France', '78.32.54.86', '78.32.54.87'),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium', 'SUPRD', 'Suprêmes délices', 'Pascale Cartrain', 'Accounting Manager', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium', '(071) 23 67 22 20', '(071) 23 67 22 21')]

In [569]:
%%sql
-- Combine order_details, products, and orders tables to get detailed order information, including the product name and quantity

SELECT *
  FROM orders AS o
  JOIN order_details AS od
    ON o.order_id = od.order_id
  JOIN products AS p
    ON p.product_id = od.product_id
 LIMIT 5;
    

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 10248, 11, 14.0, 12, 0.0, 11, 'Queso Cabrales', 5, 4, '1 kg pkg.', 21.0, 22, 30, 30, 0),
 (10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 10248, 42, 9.8, 10, 0.0, 42, 'Singaporean Hokkien Fried Mee', 20, 5, '32 - 1 kg pkgs.', 14.0, 26, 0, 0, 1),
 (10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 10248, 72, 34.8, 5, 0.0, 72, 'Mozzarella di Giovanni', 14, 4, '24 - 200 g pkgs.', 34.8, 14, 0, 0, 0),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', 10249, 14, 18.6, 9, 0.0, 14, 'Tofu', 6, 7, '40 - 100 g pkgs.', 23.25, 35, 0, 0, 0),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', 10249, 51, 42.4, 40, 0.0, 51, 'Manjimup Dried Apples', 24, 7, '50 - 300 g pkgs.', 53.0, 20, 0, 10, 0)]

In [570]:
%%sql
-- Combine employees and orders tables to see who is responsible for each order.

SELECT e.employee_id, e.first_name, e.last_name,
       o.*
  FROM orders AS o
  LEFT JOIN employees AS e
    ON o.employee_id = e.employee_id
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(5, 'Steven', 'Buchanan', 10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France'),
 (6, 'Michael', 'Suyama', 10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany'),
 (4, 'Margaret', 'Peacock', 10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil'),
 (3, 'Janet', 'Leverling', 10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France'),
 (4, 'Margaret', 'Peacock', 10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium')]

## Ranking Employee Sales Performance

Northwind Traders would like to comprehensively review the company's sales performance from an employee perspective. The objective is twofold:

- First, the management team wants to recognize and reward top-performing employees, fostering a culture of excellence within the organization.
- Second, they want to identify employees who might be struggling so that they can offer the necessary training or resources to help them improve.

In [571]:
%%sql
-- calculate total sales minus discount for each employee
WITH
employee_sales AS (
  SELECT e.employee_id, 
         e.first_name ||' '|| e.last_name AS employee_name,
         ROUND(SUM( (unit_price * quantity) - Unit_Price * Quantity * discount)::decimal,2) AS total_sales
    FROM employees AS e
    JOIN orders AS o
      ON e.employee_id = o.employee_id
    JOIN order_details AS od
      ON o.order_id = od.order_id
GROUP BY e.employee_id
)

-- use CTE to rank each employee by total sales
SELECT *,
       RANK() OVER(ORDER BY total_sales DESC) AS rank
  FROM employee_sales;

 * postgresql://postgres:***@localhost:5432/northwind
9 rows affected.


[(4, 'Margaret Peacock', Decimal('232890.85'), 1),
 (3, 'Janet Leverling', Decimal('202812.84'), 2),
 (1, 'Nancy Davolio', Decimal('192107.60'), 3),
 (2, 'Andrew Fuller', Decimal('166537.76'), 4),
 (8, 'Laura Callahan', Decimal('126862.28'), 5),
 (7, 'Robert King', Decimal('124568.23'), 6),
 (9, 'Anne Dodsworth', Decimal('77308.07'), 7),
 (6, 'Michael Suyama', Decimal('73913.13'), 8),
 (5, 'Steven Buchanan', Decimal('68792.28'), 9)]

## Running Total of Monthly Sales
Visualize the progress of the sales and identify trends that might shape the company's future strategies.

Visualize the company's sales progress over time on a monthly basis. This will involve aggregating the sales data at a monthly level and calculating a running total of sales by month. This visual will provide the management team with a clear depiction of sales trends and help identify periods of high or low sales activity.

In [572]:
%%sql
-- calculate total sales minus discount for each month
WITH
total_monthly_sales AS (
  SELECT DATE_TRUNC('month', order_date)::date AS order_month,
         ROUND(SUM( (unit_price * quantity) - Unit_Price * Quantity * discount)::decimal,2) AS total_sales
    FROM orders AS o
    JOIN order_details AS od
      ON o.order_id = od.order_id
GROUP BY order_month
)

-- calculate running total of sales by month
SELECT order_month AS "Month", total_sales AS "Total Sales",
       SUM(total_sales) OVER(ORDER BY order_month) AS "Running Total Sales"
  FROM total_monthly_sales;

 * postgresql://postgres:***@localhost:5432/northwind
23 rows affected.


[(datetime.date(1996, 7, 1), Decimal('27861.90'), Decimal('27861.90')),
 (datetime.date(1996, 8, 1), Decimal('25485.28'), Decimal('53347.18')),
 (datetime.date(1996, 9, 1), Decimal('26381.40'), Decimal('79728.58')),
 (datetime.date(1996, 10, 1), Decimal('37515.72'), Decimal('117244.30')),
 (datetime.date(1996, 11, 1), Decimal('45600.05'), Decimal('162844.35')),
 (datetime.date(1996, 12, 1), Decimal('45239.63'), Decimal('208083.98')),
 (datetime.date(1997, 1, 1), Decimal('61258.07'), Decimal('269342.05')),
 (datetime.date(1997, 2, 1), Decimal('38483.63'), Decimal('307825.68')),
 (datetime.date(1997, 3, 1), Decimal('38547.22'), Decimal('346372.90')),
 (datetime.date(1997, 4, 1), Decimal('53032.95'), Decimal('399405.85')),
 (datetime.date(1997, 5, 1), Decimal('53781.29'), Decimal('453187.14')),
 (datetime.date(1997, 6, 1), Decimal('36362.80'), Decimal('489549.94')),
 (datetime.date(1997, 7, 1), Decimal('51020.86'), Decimal('540570.80')),
 (datetime.date(1997, 8, 1), Decimal('47287.67'), Decimal('587858.47')),
 (datetime.date(1997, 9, 1), Decimal('55629.24'), Decimal('643487.71')),
 (datetime.date(1997, 10, 1), Decimal('66749.23'), Decimal('710236.94')),
 (datetime.date(1997, 11, 1), Decimal('43533.81'), Decimal('753770.75')),
 (datetime.date(1997, 12, 1), Decimal('71398.43'), Decimal('825169.18')),
 (datetime.date(1998, 1, 1), Decimal('94222.11'), Decimal('919391.29')),
 (datetime.date(1998, 2, 1), Decimal('99415.29'), Decimal('1018806.58')),
 (datetime.date(1998, 3, 1), Decimal('104854.16'), Decimal('1123660.74')),
 (datetime.date(1998, 4, 1), Decimal('123798.68'), Decimal('1247459.42')),
 (datetime.date(1998, 5, 1), Decimal('18333.63'), Decimal('1265793.05'))]

## Month-Over-Month Sales Growth

The management team would like to analyze the month-over-month sales growth rate. Understanding the rate at which sales are increasing or decreasing from month to month will help the management team identify significant trends.

### Observations

- May 1998's highest drop in sales should be ignored as this is due to a lack of data for the orders in this month, with records only covering four days.
- The highest month-over-month growth was in December 1997 with an increase of 64%, although this was after a 34.8% drop in the previous month. After this significant increase sales continued to grow by at least 5%, with a maximum of 32% for the next four months.

In [573]:
%%sql
-- calculate total sales minus discount for each month
WITH
total_monthly_sales AS (
  SELECT DATE_TRUNC('month', order_date)::date AS order_month,
         ROUND(SUM( (unit_price * quantity) - unit_price * quantity * discount)::decimal,2) AS total_sales
    FROM orders AS o
    JOIN order_details AS od
      ON o.order_id = od.order_id

GROUP BY order_month
),

-- create column with previous months sales
prev_month_sales AS (
SELECT *,
       LAG(total_sales, 1) OVER(ORDER BY order_month) AS prev_month_sales
  FROM total_monthly_sales
)

-- calculate month over month sales change
SELECT order_month AS "Month", total_sales AS "Total Sales",
       total_sales - prev_month_sales AS "# Growth Rate",
       ROUND(((total_sales - prev_month_sales) / prev_month_sales ) * 100,2) || '%' AS "% Growth Rate"
  FROM prev_month_sales

 * postgresql://postgres:***@localhost:5432/northwind
23 rows affected.


[(datetime.date(1996, 7, 1), Decimal('27861.90'), None, None),
 (datetime.date(1996, 8, 1), Decimal('25485.28'), Decimal('-2376.62'), '-8.53%'),
 (datetime.date(1996, 9, 1), Decimal('26381.40'), Decimal('896.12'), '3.52%'),
 (datetime.date(1996, 10, 1), Decimal('37515.72'), Decimal('11134.32'), '42.21%'),
 (datetime.date(1996, 11, 1), Decimal('45600.05'), Decimal('8084.33'), '21.55%'),
 (datetime.date(1996, 12, 1), Decimal('45239.63'), Decimal('-360.42'), '-0.79%'),
 (datetime.date(1997, 1, 1), Decimal('61258.07'), Decimal('16018.44'), '35.41%'),
 (datetime.date(1997, 2, 1), Decimal('38483.63'), Decimal('-22774.44'), '-37.18%'),
 (datetime.date(1997, 3, 1), Decimal('38547.22'), Decimal('63.59'), '0.17%'),
 (datetime.date(1997, 4, 1), Decimal('53032.95'), Decimal('14485.73'), '37.58%'),
 (datetime.date(1997, 5, 1), Decimal('53781.29'), Decimal('748.34'), '1.41%'),
 (datetime.date(1997, 6, 1), Decimal('36362.80'), Decimal('-17418.49'), '-32.39%'),
 (datetime.date(1997, 7, 1), Decimal('51020.86'), Decimal('14658.06'), '40.31%'),
 (datetime.date(1997, 8, 1), Decimal('47287.67'), Decimal('-3733.19'), '-7.32%'),
 (datetime.date(1997, 9, 1), Decimal('55629.24'), Decimal('8341.57'), '17.64%'),
 (datetime.date(1997, 10, 1), Decimal('66749.23'), Decimal('11119.99'), '19.99%'),
 (datetime.date(1997, 11, 1), Decimal('43533.81'), Decimal('-23215.42'), '-34.78%'),
 (datetime.date(1997, 12, 1), Decimal('71398.43'), Decimal('27864.62'), '64.01%'),
 (datetime.date(1998, 1, 1), Decimal('94222.11'), Decimal('22823.68'), '31.97%'),
 (datetime.date(1998, 2, 1), Decimal('99415.29'), Decimal('5193.18'), '5.51%'),
 (datetime.date(1998, 3, 1), Decimal('104854.16'), Decimal('5438.87'), '5.47%'),
 (datetime.date(1998, 4, 1), Decimal('123798.68'), Decimal('18944.52'), '18.07%'),
 (datetime.date(1998, 5, 1), Decimal('18333.63'), Decimal('-105465.05'), '-85.19%')]

## Identifying High-Value Customers

Management now want to identify high-value customers to whom they can offer targeted promotions and special offers, which could drive increased sales, improve customer retention, and attract new customers.

To do this, they've asked you to identify customers with above-average order values. These customers might be businesses buying in bulk or individuals purchasing high-end products.

In [574]:
%%sql

--customer identification and calculates the value of each of their orders.
WITH
customer_orders AS (
SELECT o.customer_id, o.order_id,
       ROUND(SUM( unit_price * quantity * (1 - discount))::decimal, 2) AS order_value
  FROM orders AS o
  JOIN order_details AS od
    ON o.order_id = od.order_id
GROUP BY o.customer_id, o.order_id
)

SELECT *,
       CASE
       WHEN order_value > (SELECT AVG(order_value) AS avg_value FROM customer_orders) THEN 'Above Average'
       ELSE 'Below Average'
       END AS "Average Compare"
  FROM customer_orders
 LIMIT 10;



 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[('SUPRD', 11038, Decimal('732.60'), 'Below Average'),
 ('CACTU', 10782, Decimal('12.50'), 'Below Average'),
 ('FAMIA', 10725, Decimal('287.80'), 'Below Average'),
 ('GOURL', 10423, Decimal('1020.00'), 'Below Average'),
 ('TORTU', 10518, Decimal('4150.05'), 'Above Average'),
 ('WANDK', 10356, Decimal('1106.40'), 'Below Average'),
 ('FURIB', 10963, Decimal('57.80'), 'Below Average'),
 ('WHITC', 10596, Decimal('1180.88'), 'Below Average'),
 ('ROMEY', 10282, Decimal('155.40'), 'Below Average'),
 ('QUICK', 10658, Decimal('4464.60'), 'Above Average')]

### How many orders are 'Above Average' for each customer?

In [575]:
%%sql

--customer identification and calculates the value of each of their orders.
WITH
customer_orders AS (
SELECT o.customer_id, o.order_id,
       SUM( (od.unit_price * quantity) - (od.unit_price * quantity * discount)) AS order_value
  FROM orders AS o
  JOIN order_details AS od
    ON o.order_id = od.order_id
GROUP BY o.customer_id, o.order_id
),
above_or_below_avg AS (
SELECT *,
       CASE
       WHEN order_value > (SELECT AVG(order_value) AS avg_value FROM customer_orders) THEN 'Above Average'
       ELSE 'Below Average'
       END AS "Above or Below Average"
  FROM customer_orders
)

SELECT customer_id,
       (SELECT COUNT(*)
          FROM above_or_below_avg AS a
         WHERE a.customer_id = c.customer_id AND a."Above or Below Average" = 'Above Average'
       ) AS "Orders Above Average",
       (SELECT COUNT(*)
          FROM above_or_below_avg AS a
         WHERE a.customer_id = c.customer_id AND a."Above or Below Average" <> 'Above Average'
       ) AS "Orders Below Average"
  FROM customers AS c
 LIMIT 10;


 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[('ALFKI', 0, 6),
 ('ANATR', 0, 4),
 ('ANTON', 2, 5),
 ('AROUT', 3, 10),
 ('BERGS', 5, 13),
 ('BLAUS', 0, 7),
 ('BLONP', 4, 7),
 ('BOLID', 1, 2),
 ('BONAP', 8, 9),
 ('BOTTM', 4, 10)]

## Percentage of Sales for Each Category

Provide the management team with an understanding of sales composition across different product categories. By knowing the percentage of total sales for each product category, they can gain insights into which categories drive most of the company's sales.

Note: BMP / Image column in categories causing issues with loading table so removed images with 

```UPDATE categories SET picture = null;```


In [576]:
%%sql

-- total sales for each product category
WITH
category_sales AS (
SELECT c.category_id, category_name,
       SUM( (p.unit_price * quantity) - (od.unit_price * quantity * discount)) AS total_sales
  FROM categories AS c
  JOIN products AS p
    ON c.category_id = p.category_id
  JOIN order_details AS od
    ON od.product_id = p.product_id
GROUP BY c.category_id, c.category_name
)

-- percentage of total sales for each product category
SELECT category_id AS "Category ID", category_name AS "Category",
       total_sales AS "Total Sales",
       total_sales / (SELECT SUM(total_sales) FROM category_sales) * 100 AS "Sales Percentage"
  FROM category_sales

-- Note to self: Could have used SUM(total_sales) OVER () * 100 instead of a subquery for "Sales Percentage";

 * postgresql://postgres:***@localhost:5432/northwind
8 rows affected.


[(8, 'Seafood', 138698.17620495398, 10.195420973128957),
 (7, 'Produce', 106110.97892306349, 7.8000023474916125),
 (1, 'Beverages', 290923.4796905943, 21.385193573347074),
 (5, 'Grains/Cereals', 101865.78743011087, 7.487946950899802),
 (2, 'Condiments', 114390.33561090878, 8.408601026500474),
 (4, 'Dairy Products', 252305.0841197348, 18.546433822318214),
 (6, 'Meat/Poultry', 175516.2492710619, 12.901842676711414),
 (3, 'Confections', 180586.66500423528, 13.274558629602454)]

## Top Products Per Category

The management team wants to know the top three items sold in each product category. This will allow them to identify star performers and ensure that these products are kept in stock and marketed prominently.

In [577]:
%%sql

-- total sales for each product
--WITH
--product_sales AS (
    
    
SELECT p.product_id, p.category_id product_name,
       SUM( (p.unit_price * quantity) - (od.unit_price * quantity * discount)) AS total_sales
  FROM products AS p
  JOIN order_details AS od
    ON od.product_id = p.product_id
GROUP BY p.product_id, product_name
LIMIT 10
--)

 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[(74, 7, 2836.4999990463257),
 (54, 6, 5231.987357609347),
 (29, 6, 84979.61260879139),
 (71, 4, 21400.02497317195),
 (4, 2, 9109.099985733628),
 (68, 3, 9338.999996185303),
 (34, 1, 6756.399993702769),
 (51, 7, 44035.04992691204),
 (52, 5, 3349.149998380989),
 (70, 1, 11455.649988468736)]

In [578]:
%%sql

-- total sales for each product
WITH
product_sales AS ( 
SELECT p.product_id, product_name,  p.category_id,
       SUM( (p.unit_price * quantity) - (od.unit_price * quantity * discount)) AS total_sales,
       ROW_NUMBER() OVER(
                      PARTITION BY category_id
                      ORDER BY SUM( (p.unit_price * quantity) - (od.unit_price * quantity * discount)) DESC
                    ) AS sales_ranking
  FROM products AS p
  JOIN order_details AS od
    ON od.product_id = p.product_id
GROUP BY p.product_id, product_name,  p.category_id
ORDER BY p.category_id
)
 
SELECT p.category_id, category_name, product_id, product_name, sales_ranking
  FROM product_sales AS p
  JOIN categories AS c
    ON p.category_id = c.category_id
WHERE sales_ranking <= 3

 * postgresql://postgres:***@localhost:5432/northwind
24 rows affected.


[(1, 'Beverages', 38, 'Côte de Blaye', 1),
 (1, 'Beverages', 43, 'Ipoh Coffee', 2),
 (1, 'Beverages', 2, 'Chang', 3),
 (2, 'Condiments', 63, 'Vegie-spread', 1),
 (2, 'Condiments', 61, "Sirop d'érable", 2),
 (2, 'Condiments', 65, 'Louisiana Fiery Hot Pepper Sauce', 3),
 (3, 'Confections', 62, 'Tarte au sucre', 1),
 (3, 'Confections', 20, "Sir Rodney's Marmalade", 2),
 (3, 'Confections', 26, 'Gumbär Gummibärchen', 3),
 (4, 'Dairy Products', 59, 'Raclette Courdavault', 1),
 (4, 'Dairy Products', 60, 'Camembert Pierrot', 2),
 (4, 'Dairy Products', 72, 'Mozzarella di Giovanni', 3),
 (5, 'Grains/Cereals', 56, 'Gnocchi di nonna Alice', 1),
 (5, 'Grains/Cereals', 64, 'Wimmers gute Semmelknödel', 2),
 (5, 'Grains/Cereals', 42, 'Singaporean Hokkien Fried Mee', 3),
 (6, 'Meat/Poultry', 29, 'Thüringer Rostbratwurst', 1),
 (6, 'Meat/Poultry', 17, 'Alice Mutton', 2),
 (6, 'Meat/Poultry', 53, 'Perth Pasties', 3),
 (7, 'Produce', 51, 'Manjimup Dried Apples', 1),
 (7, 'Produce', 28, 'Rössle Sauerkraut', 2),
 (7, 'Produce', 7, "Uncle Bob's Organic Dried Pears", 3),
 (8, 'Seafood', 18, 'Carnarvon Tigers', 1),
 (8, 'Seafood', 10, 'Ikura', 2),
 (8, 'Seafood', 40, 'Boston Crab Meat', 3)]